pandas/tests/groupby/test_groupby.py

```diff
@@ -120,6 +126,52 @@ def checkit(dtype):
         for dtype in ['int64', 'int32', 'float64', 'float32']:
             checkit(dtype)
 
+    def test_basic_with_nan(self):
+
+        # GH 3729
+
+        def checkit(dtype):
+            data = Series(np.arange(9) // 3, index=np.arange(9), dtype=dtype)
+
+            index = np.arange(9)
```

Comment:
move all tests to a new class, maybe TestNanGrouping (you can even put it in a new file).

We need quite comprehensive testing and it will require several tests (over multiple dtypes)